### ISSUE TITLE
playground's remote publish feature should be configurable

### ISSUE THREAD

**Issue by eMBee** [guillep/pharo-core#7123]
currently the playground remote publish feature sends the contents of the playground to a hardcoded web address.

i use the playground for private notes and i'd like to be able to publish them on my own server where i can keep them private.

i believe this can be done in several steps:

- [ ]   add a url field in the settings and let it default to the current url.

- [ ]  modify or call GTUrlProvider to use that setting. (i don't know if a configuration change could trigger a call to urlProvider: or stringUrlProvider: or if the value would need to be pulled from within GTUrlProvider)
    .
- [ ]  document the API for http://ws.stfx.eu/ and identify if there are compatible services (pastebin, etc...)
- [ ]  build a mechanism to support different APIs
- [ ]  create adapters for popular pastebin services
- [ ]  offer supported alternative services in the settings as a choice for selection
    .
- [ ]  consider disabling the feature by default and add a one-time prompt to explain how to configure and enable it.
- [ ]  consider other storage mechanisms such as git (local or remote), local files, or even a remote pharo image.

i could work on this myself (depending on how much time i can free up for this) but i am also willing to mentor a junior developer who wants to work on this. i believe this issue is suitable for a novice who has no prior experience working with smalltalk.

**Comment by estebanlm** [comment]
Hi, @eMBee, 
I don't know if you are talking about P8 or P9. In P9 we are replacing the GTPlayground (and all glamour based tools) by Spec2 tools. The new playground is in already place.
We have planned to extend the publish to cloud function, but in a complete different way you are describing (because that would mean normalize publishing in a different way and lower level than needed, IMO). Also, we target the ability to pack this different ways of publish as plugins and not to load in all cases. 
In new playground there is a `StPlaygroundPublishCommand`. Out idea is to transform that into a hierarchy, that could allow: 

- `StPlaygroundPublishToPastebinCommand`
- `StPlaygroundPublishAsGistCommand`

etc.

If you want to take a look and start there, I will be happy to help you (and I will be happy you help me to achieve this ;) )

**Comment by eMBee** [comment]
i was looking at P8 and did only a quick verification that the code was still the same in P9, but i didn't realize that the playground was replaced completely. now i see that it does indeed look noticeably different. 

i'll take a look at `StPlaygroundPublishCommand`, thank you.

does the general idea to make the default target configurable still hold?

(i can imagine another idea where the publish button on the playground offers a dropdown of available targets so the target can be chosen each time, defaulting to the previously chosen one.  that might even be preferable, because it is more flexible and allows the user to choose a different destination per playground.)

**Comment by estebanlm** [comment]
Yes, the idea is that there will be a menu button instead a single button in the case of commands with same category ("publishing", would be this one?... I'm just brainstorming here, nothing has been made in this sense). 
Anyway, you get the idea :)

**Comment by eMBee** [comment]
yes, i think the menu is a better idea than a configuration options.

so to break this down the steps would be to 

- [ ] expand `StPlaygroundPublishCommand` to multiple publish classes.
- [ ] convert the publish button into a menu showing a list of available publish options
- [ ] remember the last used publish option for this playground (maybe activate it through double click, or move the selection to the top of the list.

much simpler and more flexible than my initial idea. 

how would you go about to expand the classes?
turn `StPlaygroundPublishCommand` into an interface and move the actual implementation to subclasses?

**Comment by estebanlm** [comment]
for the menu option (remembering the last option) there is already `SpToolbarMenuButtonPresenter` (to be included in the toolbar). 
The real problem here is figure out how to map commander2 to this particular behavior.
Maybe @Ducasse or @juliendelplanque know something?

EDIT: Otherwise I have a couple of ideas, but I will not have time to work on this in some time.

**Comment by eMBee** [comment]
please share your ideas. i don't know how much time i will be able to put into this, but i have an interest to make pharo more usable for me as a desktop, and so i'll eventually want to see some progress on issues like this.
